compiler/src/iree/compiler/Preprocessing/TransformExtensions/PreprocessingExtensions.cpp

```diff
@@ -0,0 +1,397 @@
+  if (!symbolOp) {
+    return emitDefiniteFailure() << "could not find corresponding symbol op";
+  }
+  // Require isolated from above as the clone does not make sense with escaping
+  // values.
+  if (!symbolOp->hasTrait<OpTrait::IsIsolatedFromAbove>()) {
+    return emitDefiniteFailure()
+           << "target symbol op is not isolated from above";
+  }
+  StringRef symbol = getSymbol().getLeafReference();
+  SmallVector<Operation *> results;
+  for (Operation *payloadOp : state.getPayloadOps(getSymbolTable())) {
+    if (!payloadOp->hasTrait<OpTrait::SymbolTable>()) {
+      return emitDefiniteFailure()
+             << "target symbol table " << payloadOp << " is not a symbol table";
+    }
+    SymbolTable symbolTable(payloadOp);
+
+    if (Operation *preExistingSymbolOp = symbolTable.lookup(symbol)) {
+      if (getForceImport()) {
+        // If we want to overwrite pre-existing symbols, just erase it here.
+        symbolTable.erase(preExistingSymbolOp);
+      } else if (getIfUndefined()) {
+        // Skip if we want to use the symbol that is already there.
+        results.push_back(preExistingSymbolOp);
+        continue;
+      } else {
+        return emitDefiniteFailure()
+               << "target symbol " << symbol << " is already defined";
+      }
+    }
+
+    // Symbol table ops must have exactly one region with exactly one block.
+    // Simply clone the target symbol op into the single block.
+    rewriter.setInsertionPointToStart(&payloadOp->getRegion(0).front());
+    results.push_back(rewriter.clone(*symbolOp));
+  }
+  transformResults.set(cast<OpResult>(getClonedSymbol()), results);
+  return DiagnosedSilenceableFailure::success();
+}
+
+void IREE::transform_dialect::ImportSymbolOp::getEffects(
+    SmallVectorImpl<MemoryEffects::EffectInstance> &effects) {
+  transform::onlyReadsHandle(getSymbolTable(), effects);
+  transform::producesHandle(getClonedSymbol(), effects);
+  transform::modifiesPayload(effects);
+}
+
+LogicalResult IREE::transform_dialect::ImportSymbolOp::verify() {
+  if (getForceImport() && getIfUndefined()) {
+    return emitOpError()
+           << "force_import and if_undefined are mutually exclusive";
+  }
+  if (!SymbolTable::lookupNearestSymbolFrom(*this, getSymbol())) {
+    return emitOpError() << "invalid import of undefined symbol";
+  }
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// MatchCastCompatibleDagFromRootOp
+//===----------------------------------------------------------------------===//
+
+static bool isCastableToTensorType(Type from, RankedTensorType to) {
+  auto tensorType = dyn_cast<RankedTensorType>(from);
+  if (!tensorType) {
+    return false;
+  }
+  if (tensorType.getRank() != to.getRank()) {
+    return false;
+  }
+  if (tensorType.getElementType() != to.getElementType()) {
+    return false;
+  }
+  auto isAlignment = to.getEncoding()
```

Comment:
I am open to other suggestions, but I needed some way to check for static shapes with a specific alignment. For example, if I write a kernel that supports any `tensor<Axf32>` where `64 | A`, then the match should succeed for `tensor<128xf32>` or `tensor<256xf32>` but not dynamic or something else.

The encoding was just the most natural way to do this, but it doesn't really compose with any input IR that actually uses the encoding like you're noting.

Reply:
I dont think we want to support that case specifically. We can stick with dynamic or static checking for now (dynamic matching on static input as well). We maybe need better matchers to check if `tensor.dim % 64 == 0`.
Also, I was thrown off by the alignment, cause  alignment makes me think of "n"-byte alignment (which this is maybe, but not really the intent AFAICS)

Reply:
I suspect this will be a case people will want to use, especially for specialization between different shapes, but I'm fine with dropping the support here if this is not the right way to do it.

Reply:
Actually just a matcher for tensor.dim % 64 sounds good enough. I don't know why I made this so complicated...